doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1522,144 +1507,122 @@ Managing the local context
-      behaves as :tacn:`specialize`.
-
-      .. exn:: Variable @ident is already declared.
-         :undocumented:
+   .. exn:: The term "@type" has type "@type__1" which should be Set, Prop or Type.
 
-.. tacv:: eassert @type as @simple_intropattern by @tactic
-   :name: eassert
-
-   While the different variants of :tacn:`assert` expect that no existential
-   variables are generated by the tactic, :tacn:`eassert` removes this constraint.
-   This lets you avoid specifying the asserted statement completely before starting
-   to prove it.
+      Arises when the argument :token:`type` is neither of type :g:`Prop`,
+      :g:`Set` nor :g:`Type`.
 
-.. tacv:: pose proof @term {? as @simple_intropattern}
-   :name: pose proof
+   .. exn:: Proof is not complete.
+      :name: Proof is not complete. (assert)
 
-   This tactic behaves like :n:`assert @type {? as @simple_intropattern} by exact @term`
-   where :token:`type` is the type of :token:`term`. In particular,
-   :n:`pose proof @term as @ident` behaves as :n:`assert (@ident := @term)`
-   and :n:`pose proof @term as @simple_intropattern` is the same as applying the
-   :token:`simple_intropattern` to :token:`term`.
+      :n:`@ltac_expr3` was not able to prove the provided :n:`@term` or :n:`@one_term`.
 
-.. tacv:: epose proof @term {? as @simple_intropattern}
-   :name: epose proof
+   .. tacn:: eassert ( @ident : @type ) {? by @ltac_expr3 }
+             eassert @one_term {? @as_ipat } {? by @ltac_expr3 }
+      :name: eassert; _
 
-   While :tacn:`pose proof` expects that no existential variables are generated by
-   the tactic, :tacn:`epose proof` removes this constraint.
+      .. todo PR Need a better spot to define and comment on holes.
+         Maybe that could be in the existential variables chapter,
+         but making clear that they are existential variables?
 
-.. tacv:: pose proof (@ident := @term)
+      Unlike :tacn:`assert`, the :n:`@type`\s and :n:`@one_term`\s in
+      :tacn:`eassert` may contain :gdef:`holes <hole>`, denoted by :n:`_`,
+      for which the tactic will create existential variables.  This lets you
+      avoid specifying the asserted statement completely before starting to
+      prove it.
 
-   This is an alternative syntax for :n:`assert (@ident := @term)` and
-   :n:`pose proof @term as @ident`, following the model of :n:`pose
-   (@ident := @term)` but dropping the value of :token:`ident`.
+.. tacn:: enough ( @ident : @type ) {? by @ltac_expr3 }
+          enough @one_term {? @as_ipat } {? by @ltac_expr3 }
+   :name: enough; _
 
-.. tacv:: epose proof (@ident := @term)
+   .. todo PR The only difference vs assert is where the goal is added?  If so, then
+      this entry should only refer to assert and mention the difference(s).
+      There are several similar tactics in this section.  Maybe we should add a
+      summary of their differences?  Perhaps you could briefly outline them?
```

Comment:
The location of the goal when `by @ltac_expr3` is not provided, and how this clause applies when it is provided. (And also there is no `(@ident := @term)` form here.)  I honestly think that your current description is pretty good and already relates to `assert` sufficiently.

Reply:
I strongly prefer describing things in terms of differences rather than repeating similar text.  Less to read and wonder whether minor wording differences are significant or not.  No need to mention that assert has one more form, that should be pretty apparent.

The part about "possibly destructed" also applies to `assert`?

Reply:
By the way, this seems like a bug:

```
enough (True /\ True) by split; apply I.
(*
in Tactics.check_number_of_constructors:
Not an inductive goal with 1 constructor.
*)
```

Reply:
> The part about "possibly destructed" also applies to `assert`?

It doesn't because `ltac_expr3` and the intropattern are not applied to the same subgoal in the case of `assert` (contrary to `enough`).

Since you want to describe them in terms of difference, here is my summary of the differences:

- The simple case is when none of `ltac_expr3` and `as_ipat` are provided. In this case, the tactics both create a subgoal corresponding to the initial goal with the added hypothesis and a subgoal asking to prove the hypothesis. The only difference is the order (`assert` asks to prove the hypothesis first, and `enough` last).
- When `as_ipat` is provided, it is applied to the initial goal with the added hypothesis. But some patterns may create multiple subgoals, so in this case, the tactics may leave more than 2 subgoals. Only one of them corresponds to the hypothesis to prove, the rest of the subgoals will correspond to the initial goal with the added hypothesis after its possible transformations.
- When `by @ltac_expr3` is provided, it applies to the subgoal(s) that the tactic puts first, so that's the (unique) subgoal asking to prove the hypothesis in the case of `assert` and the (possibly several) subgoal(s) corresponding to the initial goal where the hypothesis has been added (and possibly transformed) in the case of `enough`. Given that this close requires to solve entirely the subgoal(s) on which it is applied, a consequence is that `enough ... as ... by ...` will never leave more than one subgoal, whereas `assert ... as ... by ...` may leave several.

> By the way, this seems like a bug:
> 
> ```
> enough (True /\ True) by split; apply I.
> (*
> in Tactics.check_number_of_constructors:
> Not an inductive goal with 1 constructor.
> *)
> ```

I cannot reproduce your exact error message without knowing the context, but what you observe is probably a parenthesis issue: you need to put parentheses around `split; apply I`. This is why this is an `ltac_expr3`.

Reply:
If it has 6 legs, it's a bug :-)
```
Goal False.
enough (True /\ True) by (split; apply I).
(*
in Tactics.check_number_of_constructors:
Not an inductive goal with 1 constructor.
*)
```

However this works: `assert (True /\ True) by (split; apply I).` as does `enough (0 = 0 /\ 1 = 1) by (split; reflexivity).`

Reply:
Which one does "It's" refer to?  See Zulip.

Reply:
The "by" clause should apply to the goals created by assert/enough, not the entire context.  Otherwise, it adds no functionality: you could split off the by clause as a separate right after the assert/enough.

Reply:
If you look in `g_tactic.mlg`, assert, enough, pose and their e* variants all invoke `TacAssert` with varying parameters.  The second parameter is false only for [e]enough, not for the others.  The code for processing the tactic is in `tacinterp.ml` (`| TacAssert (ev,b,t,ipat,c) ->`, but it seems to recursively refer to TacAssert and for some reason I can't find the ultimate definition.  So can't find the code to handle that second parameter.

Reply:
> Which one does "It's" refer to? See Zulip.

It refers to "the same things".

>The "by" clause should apply to the goals created by assert/enough, not the entire context.

A subset of the goals yes. I don't get what you are asking.

> If you look in `g_tactic.mlg`, assert, enough, pose and their e* variants all invoke `TacAssert` with varying parameters. The second parameter is false only for [e]enough, not for the others. The code for processing the tactic is in `tacinterp.ml` (`| TacAssert (ev,b,t,ipat,c) ->`, but it seems to recursively refer to TacAssert and for some reason I can't find the ultimate definition. So can't find the code to handle that second parameter.

For a tactic that I use a lot and whose behavior I know and understand pretty well, I'm a bit reluctant to spend time figuring out how the code works TBH.

Reply:
> It's behaving normally.

I think not, but let's consider these simpler examples instead.  In the first one, the `by` clause applies only to the new goal `True`, as I would expect.  In the second, the `by` clause is applied to (at least) the original goal `False`, which seems like a bug to me.  IIUC the point of having the `by` clause is to apply it only to (some of?) the goal given in the tactic.  If not, how do I describe this behavior in the doc?

```
Goal False. assert True by apply I. (* works, still have "False" goal *)
Goal False. enough True by apply I. (* Unable to unify "True" with "False". *)

Reply:
Or are you saying that `enough one_term by tactic.` is equivalent to `enough one_term. tactic.`, which means that the `by` doesn't add any functionality?

Reply:
> Or are you saying that `enough one_term by tactic.` is equivalent to `enough one_term. tactic.`, which means that the `by` doesn't add any functionality?

Precisely.

Reply:
`by` never adds any functionality AFAICT, it is always a way of clarifying that you can solve the side-goals that are generated with the provided tactic. So it makes for a much clearer and readable script.

Reply:
And obviously `enough one_term as [|] by tactic.` is not equivalent to `enough one_term as [|]. tactic.` but as `enough one_term as [|]. 1-2: tactic.`

Reply:
I disagree with calling that "no added functionality" BTW, otherwise where is the limit where you consider something a functionality? I could push your reasoning and say that no tactic is useful beyond `exact` since I can prove everything by providing the proof term directly.

Reply:
> IIUC the point of having the `by` clause is to apply it only to the goal given in the tactic. If not, how do I describe this behavior in the doc?

The point of the `by` clause is to discharge side-goals. The main difference between `assert` and `enough` is that the user does not consider the side-goal to be the same. That's a choice that the user makes when writing the proof script. The side-goal is shown first and it will be sometimes put inside focusing braces, or if it can be solved quickly a `by` clause will be used.

Reply:
> I disagree with calling that "no added functionality" BTW

I think you're over analyzing a simple observation.